src/inmanta/deploy/state.py

```diff
@@ -142,28 +202,110 @@ class ModelState:
     """
 
     version: int
-    resources: dict[ResourceIdStr, ResourceDetails] = dataclasses.field(default_factory=dict)
+    resources: dict["ResourceIdStr", ResourceDetails] = dataclasses.field(default_factory=dict)
     requires: RequiresProvidesMapping = dataclasses.field(default_factory=RequiresProvidesMapping)
-    resource_state: dict[ResourceIdStr, ResourceState] = dataclasses.field(default_factory=dict)
+    resource_state: dict["ResourceIdStr", ResourceState] = dataclasses.field(default_factory=dict)
     # resources with a known or assumed difference between intent and actual state
     # (might be simply a change of its dependencies), which are still being processed by
     # the resource scheduler. This is a short-lived transient state, used for internal concurrency control. Kept separate from
     # ResourceStatus so that it lives outside the scheduler lock's scope.
-    dirty: set[ResourceIdStr] = dataclasses.field(default_factory=set)
+    dirty: set["ResourceIdStr"] = dataclasses.field(default_factory=set)
     # types per agent keeps track of which resource types live on which agent by doing a reference count
     # the dict is agent_name -> resource_type -> resource_count
-    types_per_agent: dict[str, dict[ResourceType, int]] = dataclasses.field(
+    types_per_agent: dict[str, dict["ResourceType", int]] = dataclasses.field(
         default_factory=lambda: defaultdict(lambda: defaultdict(lambda: 0))
     )
 
+    @classmethod
+    async def create_from_db(
```

Comment:
This one scares mea bit, but I may not be seeing all ends. 

1. I does some rather delicate operations, manipulating a complex assembly of fields.
I would expect this kind of logic to be more isolated, but admittedly, every other method in this class is like this. 
There is no central 'update_state' method, instead, this entire class contains specific update methods. 
This may be for good reason, but it struct me as particular. 


2. For testing, it might be easier to split it in a DB related method that builds resource_state and details and the one that installs it. This way, the delicate loading logic can be tested without any DB related thing. (it could be less memory efficient, but I doubt it, as we load it all into memory )